x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.searchablesnapshots;
+
+import org.elasticsearch.cluster.routing.RecoverySource;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+import org.elasticsearch.cluster.routing.allocation.AllocateUnassignedDecision;
+import org.elasticsearch.cluster.routing.allocation.AllocationDecision;
+import org.elasticsearch.cluster.routing.allocation.ExistingShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.FailedShard;
+import org.elasticsearch.cluster.routing.allocation.NodeAllocationResult;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SearchableSnapshotAllocator implements ExistingShardsAllocator {
+
+    static final String ALLOCATOR_NAME = "searchable_snapshot_allocator";
+
+    @Override
+    public void allocateUnassigned(RoutingAllocation allocation) {
+
+        final RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
+
+        while (iterator.hasNext()) {
```

Comment:
It seems like the loop here and the handling of a NO decision belongs in the framework (allocation service?) rather than in every plugin. It is important that this is done right.

This in turn could lead to removing `GatewayAllocator` completely and letting Replica and PrimaryShardAllocator become individual `ExistingShardsAllocator`s. A method to allow `processExistingRecoveries` would be necessary though.

Reply:
Here we handle `NO` (and `THROTTLE`) decisions by stopping further allocation attempts, but that behaviour seems to me to be specific to this plugin (and the fact that this naive implementation checks all nodes against the deciders). For instance, the `ReplicaShardAllocator` does not share this behaviour: it only checks the deciders for nodes with at least some of the shard on disk already, and allows `NO` decisions to fall through to the `BalancedShardsAllocator` so that replicas can be allocated afresh.

I'm in two minds about iterating through the unassigned shards within each allocator. As written, we prioritise the shards that can be allocated by a plugin over the standard shards. In the case of searchable snapshots I don't think this is too bad (they should be fast to start and may well live on their own nodes). On the other hand if we pulled the loop up to the `AllocationService` then this would change the allocation order to prefer recovering higher-priority replicas instead of lower-priority primaries, which would keep the cluster in `red` health for longer after startup.

Reply:
Hmm having written that it's now obvious that we could solve the priority thing by sorting the shards based on whether they're a primary or not 🤦‍♂